library/src/blas2/rocblas_tpsv.hpp

```diff
@@ -0,0 +1,304 @@
+/* ************************************************************************
+ * Copyright 2016-2020 Advanced Micro Devices, Inc.
+ * ************************************************************************ */
+
+#include "../blas1/rocblas_copy.hpp"
+#include "handle.h"
+#include "logging.h"
+#include "rocblas.h"
+#include "utility.h"
+#include <algorithm>
+#include <cstdio>
+#include <tuple>
+
+namespace
+{
+    rocblas_int
+        packed_matrix_index(bool upper, bool trans, rocblas_int n, rocblas_int row, rocblas_int col)
+    {
+        return upper ? (trans ? ((row * (row + 1) / 2) + col) : ((col * (col + 1) / 2) + row))
+                     : (trans ? (((row * (2 * n - row + 1)) / 2) + (col - row))
+                              : (((col * (2 * n - col + 1)) / 2) + (row - col)));
+    }
+
+    // Uses forward substitution to solve Ax = b. Used for a non-transposed lower-triangular matrix
+    // or a transposed upper-triangular matrix.
+    template <bool CONJ, rocblas_int BLK_SIZE, typename T>
+    __device__ void tpsv_forward_substitution_calc(
+        bool diag, bool trans, int n, const T* A, T* x, rocblas_int incx)
+    {
+        __shared__ T xshared[BLK_SIZE];
+        int          tx = threadIdx.x;
+
+        // main loop - iterate forward in BLK_SIZE chunks
+        for(rocblas_int i = 0; i < n; i += BLK_SIZE)
+        {
+            // cache x into shared memory
+            if(tx + i < n)
+                xshared[tx] = x[(tx + i) * incx];
+
+            __syncthreads();
+
+            // iterate through the current block and solve elements
+            for(rocblas_int j = 0; j < BLK_SIZE; j++)
+            {
+                // If the current block covers more than what's left,
+                // we break early.
+                if(j + i >= n)
+                    break;
+
+                // solve element that can be solved
+                if(tx == j && !diag)
+                {
+                    rocblas_int colA   = j + i;
+                    rocblas_int rowA   = j + i;
+                    rocblas_int indexA = packed_matrix_index(trans, trans, n, rowA, colA);
+                    xshared[tx]        = xshared[tx] / (CONJ ? conj(A[indexA]) : A[indexA]);
+                }
+
+                __syncthreads();
```

Comment:
Are these syncthreads safe? I think that all threads in a warp must be able to reach the syncthread, however if some threads here hit your break statement and some dont (within in the same warp) I dont know if this is safe or not. Maybe double check with torre if you dont know. See also: https://stackoverflow.com/questions/6666382/can-i-use-syncthreads-after-having-dropped-threads

Reply:
I remember seeing some sort of documentation on this somewhere, I'll try to find it again as a reference. If only the documentation was easily accessible...
It seems that what Torre said stands, I haven't tried anything more complicated than what's done here, not sure how `if`/`else` statements would be handled.

Reply:
I couldn't find any documentation :( . For now I changed it to conform to the rules set by cuda. Can take another look if you want to make sure.

Reply:
OK sounds good. As a side test it would be interesting to try the case where BLK_SIZE is the warp size (I know its 32 for cuda but maybe its 64 for hip?). In this case I dont think you need the syncthreads as long as you make the shared memory volatile. Be interesting to see if its faster.